### ISSUE TITLE
Cilium Endpoints are not reachable between cluster (Master/Worker Nodes) using autoDirectNodeRoutes=true 

### ISSUE THREAD

**Issue by hemantbali** [cilium/cilium#28302]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

 i have k8s clusters (kubeadm + ubuntu 20.04), i have installed cilium using below helm. But the endpoint is not reachable from master node to worker node and visa versa, However both nodes can reach each other because they are in same subnet (192.168.20.0/24).
i am running native routing mode and have used autoDirectNodeRoutes=true, so that cilium can inject the routes automatically. And cilium is taking care of the routes.

helm install cilium cilium/cilium --version 1.14.1 --namespace kube-system \
  --set tunnel=disabled \
  --set routingMode=native \
  --set kubeProxyReplacement=true \
  --set ipv4.enabled=true \
  --set ipam.operator.clusterPoolIPv4PodCIDRList="10.1.0.0/16" \
  --set ipv4NativeRoutingCIDR="10.0.0.0/8" \
  --set bpf.masquerade=true \
  --set bpf.lbExternalClusterIP=true \
  --set autoDirectNodeRoutes=true \
  --set k8sServiceHost=ku-master-1.example.net \
  --set k8sServicePort=6443 \
  --set enableIPv4Masquerade=true \
  --set cluster.name=cluster1 \
  --set cluster.id=1

# Cilium agents on both the nodes are up.
root@ku-master-1:/home/ubuntu# kubectl -n kube-system get pods -l k8s-app=cilium -o wide
NAME           READY   STATUS    RESTARTS   AGE   IP               NODE                      NOMINATED NODE   READINESS GATES
cilium-4xwvt   1/1     Running   0          23m   192.168.20.115   ku-worker-1.example.net   <none>           <none>
cilium-8n45h   1/1     Running   0          23m   192.168.20.39    ku-master-1.example.net   <none>           <none>

# checking the cilium health from cilium agent in the master node.
root@ku-master-1:/home/ubuntu# kubectl -n kube-system -ti exec cilium-8n45h -- cilium-health status
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
Probe time:   2023-09-26T12:20:32Z
Nodes:
  cluster1/ku-master-1.example.net (localhost):
    Host connectivity to 192.168.20.39:
      ICMP to stack:   OK, RTT=302.387µs
      HTTP to agent:   OK, RTT=427.163µs
    Endpoint connectivity to 10.1.1.6:
      ICMP to stack:   OK, RTT=290.972µs
      HTTP to agent:   OK, RTT=425.466µs
  cluster1/ku-worker-1.example.net:
    Host connectivity to 192.168.20.115:
      ICMP to stack:   OK, RTT=1.2401ms
      HTTP to agent:   OK, RTT=1.415954ms
    Endpoint connectivity to 10.1.0.196:
      ICMP to stack:   Connection timed out
      HTTP to agent:   Get "http://10.1.0.196:4240/hello": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

root@ku-master-1:/home/ubuntu# kubectl -n kube-system -ti exec cilium-8n45h -- ip route
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
default via 192.168.20.1 dev ens3 proto dhcp src 192.168.20.39 metric 100
10.1.0.0/24 via 192.168.20.115 dev ens3 proto kernel
10.1.1.0/24 via 10.1.1.130 dev cilium_host proto kernel src 10.1.1.130
10.1.1.130 dev cilium_host proto kernel scope link
169.254.169.254 via 192.168.20.2 dev ens3 proto dhcp src 192.168.20.39 metric 100
192.168.20.0/24 dev ens3 proto kernel scope link src 192.168.20.39

# checking the cilium health from cilium agent in the worker node.

root@ku-master-1:/home/ubuntu# kubectl -n kube-system -ti exec cilium-4xwvt -- cilium-health status
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
Probe time:   2023-09-26T12:27:19Z
Nodes:
  cluster1/ku-worker-1.example.net (localhost):
    Host connectivity to 192.168.20.115:
      ICMP to stack:   OK, RTT=227.742µs
      HTTP to agent:   OK, RTT=486.278µs
    Endpoint connectivity to 10.1.0.196:
      ICMP to stack:   OK, RTT=237.517µs
      HTTP to agent:   OK, RTT=1.287003ms
  cluster1/ku-master-1.example.net:
    Host connectivity to 192.168.20.39:
      ICMP to stack:   OK, RTT=1.148807ms
      HTTP to agent:   OK, RTT=1.242094ms
    Endpoint connectivity to 10.1.1.6:
      ICMP to stack:   Connection timed out
      HTTP to agent:   Get "http://10.1.1.6:4240/hello": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

root@ku-master-1:/home/ubuntu# kubectl -n kube-system -ti exec cilium-4xwvt -- ip route
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
default via 192.168.20.1 dev ens3 proto dhcp src 192.168.20.115 metric 100
10.1.0.0/24 via 10.1.0.198 dev cilium_host proto kernel src 10.1.0.198
10.1.0.198 dev cilium_host proto kernel scope link
10.1.1.0/24 via 192.168.20.39 dev ens3 proto kernel
169.254.169.254 via 192.168.20.2 dev ens3 proto dhcp src 192.168.20.115 metric 100
192.168.20.0/24 dev ens3 proto kernel scope link src 192.168.20.115

### Cilium Version

root@ku-master-1:/home/ubuntu# cilium version
cilium-cli: v0.15.8 compiled with go1.21.0 on linux/amd64
cilium image (default): v1.14.1
cilium image (stable): v1.14.2
cilium image (running): 1.14.1

### Kernel Version

root@ku-master-1:/home/ubuntu# uname -r
5.4.0-050400-generic

root@ku-master-1:/home/ubuntu# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

### Kubernetes Version

root@ku-master-1:/home/ubuntu# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"28", GitVersion:"v1.28.1", GitCommit:"8dc49c4b984b897d423aab4971090e1879eb4f23", GitTreeState:"clean", BuildDate:"2023-08-24T11:21:51Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

[cilium-sysdump-20230927-033135.zip](https://github.com/cilium/cilium/files/12734326/cilium-sysdump-20230927-033135.zip)

### Relevant log output

```shell
root@ku-master-1:/home/ubuntu# kubectl get pods -n kube-system
NAME                                              READY   STATUS    RESTARTS         AGE
cilium-4xwvt                                      1/1     Running   0                15h
cilium-8n45h                                      1/1     Running   0                15h
cilium-operator-6d95d5cbbc-pfhmt                  1/1     Running   1 (10h ago)      15h
cilium-operator-6d95d5cbbc-qkjmp                  1/1     Running   3 (7h12m ago)    15h
coredns-5dd5756b68-b4jnf                          1/1     Running   0                15h
coredns-5dd5756b68-lnff5                          1/1     Running   0                15h
etcd-ku-master-1.example.net                      1/1     Running   10               16h
kube-apiserver-ku-master-1.example.net            1/1     Running   13               15h
kube-controller-manager-ku-master-1.example.net   1/1     Running   32 (7h13m ago)   15h
kube-scheduler-ku-master-1.example.net            1/1     Running   23 (7h12m ago)   16h
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ti-mo** [comment]
Hi @hemantbali, what else have you done troubleshooting-wise? Have you confirmed (using tcpdump) packets are leaving the node?

**Comment by hemantbali** [comment]
@ti-mo , Yes the Both (Master + worker) Nodes are in same Subnet and reachable.
But problem comes with the POD CIDR, 

# Shows that the cilium agent in the worker node could not able to reach the endpoint in the master node. (10.1.1.6 is the object on the master node.)
![image](https://github.com/cilium/cilium/assets/146154975/5e1f45d8-f6bd-4a31-aa4d-f8ceb9e1e801)

# below is the routing done by cilium on the worker node.
![image](https://github.com/cilium/cilium/assets/146154975/0309bf77-a6a3-4760-bcde-43220667b8e3)

# When i run the tcp dump on the worker node to see what happens when i ping from worker node to the cilium endpoint in the master node 10.1.1.6. i can see the ICMP echo req has no reply. which means that the reachability issue. But as i said both the nodes are in same subnet and routes are their for the pods. i am not sure why it is happening ?  Here, something is missing from pod_to_Node routing map.
![image](https://github.com/cilium/cilium/assets/146154975/6e449d2e-d518-4794-8db9-6c6c8230b873)

**Comment by ti-mo** [comment]
I see there's a route for `169.254.169.254`, so I assume you're running on AWS? Did you [disable source/destination checks](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_NAT_Instance.html#EIP_Disable_SrcDestCheck) on your instances? Are you following one of our user guides for the particular platform you're targeting? I'd recommend picking a getting started guide and branching out from there.

**Comment by hemantbali** [comment]
@ti-mo , No i have these instance in my own Openstack env. yes i have created a deafult security group and attached to all the instance and that sec_grp have all the rules like ICMP, SSH and etc allowed with 0.0.0.0/0.

**Comment by hemantbali** [comment]
ti-mo, but here is the catch :) The Nodes are under same L2 Domain. From Node to Node i can see ping is working. The issue is when the **endpoint 10.1.1.6** > 192.168.20.115 unicast the ICMP Echo Reply messages why it is not on the Node 192.168.20.115. As both Nodes are on same network. I understand this doesn't look like a cilium issue. But if you could help to resolve it will be helpful.

**Comment by eviljanca** [comment]
@hemantbali were u able to resolve the issue? i'm facing exactly the same issue

**Comment by CallisteH** [comment]
Same issue here. No news ?

**Comment by CallisteH** [comment]
I finally found the solution to my networking problem.

My Kubernetes cluster was modifying the iptables rules on my nodes. When I deleted Kubernetes resources such as MetalLB, Cilium, or others, these rules didn't seem to be removed from the iptables. I'm not sure if this is expected behavior from Kubernetes or not.

To resolve this, I decided to flush my iptables and rebuild them without the unnecessary rules. Here's the process I followed:

1. First, I saved the current iptables rules to a file:

   ```
   iptables-save > my-iptables.v4
   ```

2. Then, I edited this file to remove all Cilium-related rules.

3. Finally, I reloaded the modified rules:

   ```
   iptables-restore < my-iptables.v4
   ```
This process effectively cleaned up my iptables rules, removing the lingering entries from deleted Kubernetes resources.

Then, I reinstall the CNI Cilium. And no errors appear.
